### PR TITLE
Fix List operation bulk buttons on XS and SM

### LINKED ISSUE #3135: [4.1][Bug] Mobile listOperation does not show the Export button
# Feature Request

### What's the feature you think Backpack should have?

From a mobile, the Export button is hidden, this happens when the "BulkDeleteOperation" option is enabled.

Screenshot_20200815-200047_Chrome

Another semanagement is that the header that comes out at the bottom can be disabled, since it remains at the top in modal mode.

If the "disableResponsiveTable ()" function is enabled, the top header is not output in modal mode.

### PR DESCRIPTION
**Pull request by tabacitu**
Fixes #3135 
Alternative to #3181
Merges #3205 

I've inspected the code a little bit, and it looks like our intention was never to show the Export buttons on screens this small. 

In the previous version (Backpack 3.x), we used `hidden-xs` to hide the export buttons on mobile entirely. And it makes sense, why would you hide/show columns when so few columns are visible. I still think that's a good idea.

So this PR:
- hides the export buttons on XS
- fixes a few print issues (looks like we hadn't replaced Bootstrap 3's `hidden` print with Bootstrap 4's `d-print-none`)
- centers the bulk buttons on mobile, so it looks a little better
- fixes the problem using helper classes in CSS, so that we can push it as a non-breaking change

Here's the end result:

@pxpm & @DanielKimmich - I'd love it if you guys could give it a go, and let me know if you discover something I've missed. Anybody else is welcome to test too, of course. _I think_ I've covered all use cases, but maybe I'm wrong:
- test by doing  `composer require backpack/crud:"dev-fix-operation-display-on-print-xs-and-sm"` and a hard refresh;
- go back to normal with `composer require backpack/crud:"4.1.*"`

Cheers!

### CODE REVIEW

**Comment by pxpm** [comment]
I think in the end we can not make everybody happy and have to make some decision. 

I personally would like those buttons on page, like you said, we probably can show 1/2 columns not more, it would be **great** if I can show the columns I want.

Imagine:
| id | car | manufacturer | plate | status | owner |
|----|-----|--------------|-------|--------|-------|

My guess is that probably `id` and `car` would be visible, but I am looking for `status`, I even searched by `status`, and my results are filtered .. if I could just see plate or owner ..

Ofcouse, I can try to turn my phone (hope it's not locked rotation), if so, I must know how to go to settings and enable screen rotation. 

I am trying to put myself in some random enduser that happens to use the software that some of our community devs created.

From developer perspective removing the buttons might be a wise decision, I personally have nothing against, I happen to know that I can turn my screen and how to activate screen rotation (yes, I am one of those that has screen rotation locked, that's why that came to mind).

What I notice is that some apps that I use force `landscape mode`, so I am good with that. 

I tried searching about this subject, from my research I found that yes, is possible, and no it's not to do it from html.

For reference: 
https://stackoverflow.com/questions/14360581/force-landscape-orientation-mode
http://fuel-efficient-vehicles.org/pwsdb/?p=330

But I am not also sure that forcing landscape is the way to go. 

To sum up, I will test this, I am ok removing the buttons, but I still have the concerns I pointed out. 

Best,
Pedro

**Comment by tabacitu** [comment]
I understand your concerns @pxpm , thank you. Glad we had the talk. I've worked on this some more, and merged the results into https://github.com/Laravel-Backpack/CRUD/pull/3205 so let's close this in favour of that.

**Comment by tabacitu** [comment]
Thank you for voicing your concerns @pxpm . I think you're right, it does kind of make sense to pick which columns to show, when you can only show so few columns.

So I've tried to do it. A lot. But no matter what I did, I couldn't keep the export buttons on XS _and_ make them look decent _and_ make it a reasonably non-breaking change (without major changes in `list.css`). So in the end I've decided to hide it on XS, under the reasoning that people can (and should) switch to landscape to see more functionality. I'm not 100% happy with it either, but we can change this behaviour in 4.2 when we can make more changes to `list.css`. Until then, it's better to hide the buttons than keep them looking odd.

In addition to the above, I've been able to squeeze into this PR quite a few more improvements to the XS, SM and print user experience. As much as it's possible without considering it a breaking change, I think:
- hides the export buttons on XS;
- makes the print view cleaner (no buttons or actions);
- centers the bulk buttons on mobile, so it looks a little better;
- fixes the problem using helper classes in CSS, so that we can push it as a non-breaking change;
- changes the default priorities given to columns, so that the `Actions` column (the one with the Edit/Delete buttons) does NOT show up on XS and SM, since it's usually large and doesn't leave room for an identifiable column to show; by default:
   - the first column added is given priority 0;
   - the bulk action columns are given priority 0;
   - the action column (Edit/Delete) keeps its default priority 1 (which means it's less important than the above);
- it does NOT do Laravel-Backpack/ideas#70 yet (does not hide bulk actions on mobile) because I found no reasonably way to do it; so the problem remains that when using `bulkActions` + `responsiveTable` the only column that gets shown is the freakin checkbox. I've spent >3 hours trying to fix that alone, no luck. Will try again but until then some fixes are better than no fixes;

The end result, I think, is decent, while still having problem Laravel-Backpack/ideas#70  which hopefully we _can_ and _will_ solve soon:

File: src/public/packages/backpack/crud/css/list.css [hunk 1]
```diff
@@ -54,7 +54,8 @@
 	}
 
   /*  DataTables Loading State Visual Improvements  */
-
+	table.dataTable.dtr-inline.collapsed>tbody>tr[role="row"]>td.dtr-control:before, 
+	table.dataTable.dtr-inline.collapsed>tbody>tr[role="row"]>th.dtr-control:before,
```
Comment (tabacitu) [inline src/public/packages/backpack/crud/css/list.css:58]: This is to future-proof it, it's not _actually_ needed, it will be needed in 4.2 when we get to actually hide the BulkActions column on XS.

So it's not a breaking change, even if it's a change to `list.css`

File: src/app/Library/CrudPanel/Traits/Read.php [hunk 2]
```diff
@@ -143,26 +143,28 @@ public function enableBulkActions()
...
             'label'           => ' ',
-            'priority'        => 1,
+            'priority'        => 0,
             'searchLogic'     => false,
             'orderable'       => false,
             'visibleInTabel'  => true,
             'visibleInModal'  => false,
             'visibleInExport' => false,
             'visibleInShow'   => false,
+            'hasActions'      => true,
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Read.php:167]: By adding a `hasActions` flag to columns, we can easily count how many _real_ columns there are, and exclude the columns that have actions.

File: src/resources/views/crud/create.blade.php [hunk 3]
```diff
@@ -18,7 +18,7 @@
         <small>{!! $crud->getSubheading() ?? trans('backpack::crud.add').' '.$crud->entity_name !!}.</small>
 
         @if ($crud->hasAccess('list'))
-          <small><a href="{{ url($crud->route) }}" class="hidden-print font-sm"><i class="la la-angle-double-{{ config('backpack.base.html_direction') == 'rtl' ? 'right' : 'left' }}"></i> {{ trans('backpack::crud.back_to_all') }} <span>{{ $crud->entity_name_plural }}</span></a></small>
+          <small><a href="{{ url($crud->route) }}" class="d-print-none font-sm"><i class="la la-angle-double-{{ config('backpack.base.html_direction') == 'rtl' ? 'right' : 'left' }}"></i> {{ trans('backpack::crud.back_to_all') }} <span>{{ $crud->entity_name_plural }}</span></a></small>
```
Comment (tabacitu) [inline src/resources/views/crud/create.blade.php:21]: Just `print` fixes - looks like we didn't change the classes when moving from Bootstrap 3 to Bootstrap 4.

File: src/resources/views/crud/edit.blade.php [hunk 4]
```diff
@@ -18,7 +18,7 @@
         <small>{!! $crud->getSubheading() ?? trans('backpack::crud.edit').' '.$crud->entity_name !!}.</small>
 
         @if ($crud->hasAccess('list'))
-          <small><a href="{{ url($crud->route) }}" class="hidden-print font-sm"><i class="la la-angle-double-{{ config('backpack.base.html_direction') == 'rtl' ? 'right' : 'left' }}"></i> {{ trans('backpack::crud.back_to_all') }} <span>{{ $crud->entity_name_plural }}</span></a></small>
+          <small><a href="{{ url($crud->route) }}" class="d-print-none font-sm"><i class="la la-angle-double-{{ config('backpack.base.html_direction') == 'rtl' ? 'right' : 'left' }}"></i> {{ trans('backpack::crud.back_to_all') }} <span>{{ $crud->entity_name_plural }}</span></a></small>
```
Comment (tabacitu) [inline src/resources/views/crud/edit.blade.php:21]: Just `print` fixes - looks like we didn't change the classes when moving from Bootstrap 3 to Bootstrap 4.

File: src/resources/views/crud/inc/datatables_logic.blade.php [hunk 5]
```diff
@@ -209,9 +209,9 @@ functionsToRunOnDataTablesDrawEvent: [],
               "type": "POST"
           },
           dom:
-            "<'row hidden'<'col-sm-6 hidden-xs'i><'col-sm-6 hidden-print'f>>" +
+            "<'row hidden'<'col-sm-6'i><'col-sm-6 d-print-none'f>>" +
```
Comment (tabacitu) [inline src/resources/views/crud/inc/datatables_logic.blade.php:212]: Just `print` fixes - looks like we didn't change the classes when moving from Bootstrap 3 to Bootstrap 4.

File: src/resources/views/crud/inc/datatables_logic.blade.php [hunk 6]
```diff
@@ -209,9 +209,9 @@ functionsToRunOnDataTablesDrawEvent: [],
...
             "<'row'<'col-sm-12'tr>>" +
-            "<'row mt-2 '<'col-sm-6 col-md-4'l><'col-sm-2 col-md-4 text-center'B><'col-sm-6 col-md-4 hidden-print'p>>",
+            "<'row mt-2 d-print-none '<'col-sm-12 col-md-4'l><'col-sm-0 col-md-4 text-center'B><'col-sm-12 col-md-4 'p>>",
```
Comment (tabacitu) [inline src/resources/views/crud/inc/datatables_logic.blade.php:214]: Includes `print` fixes - looks like we didn't change the classes when moving from Bootstrap 3 to Bootstrap 4.

Also makes the bulk buttons show up nicely on XS. Centered.

File: src/resources/views/crud/list.blade.php [hunk 7]
```diff
@@ -30,15 +30,15 @@
...
+              <div class="d-print-none {{ $crud->hasAccess('create')?'with-border':'' }}">
 
                 @include('crud::inc.button_stack', ['stack' => 'top'])
 
               </div>
             @endif
           </div>
           <div class="col-sm-6">
-            <div id="datatable_search_stack" class="mt-sm-0 mt-2"></div>
+            <div id="datatable_search_stack" class="mt-sm-0 mt-2 d-print-none"></div>
```
Comment (tabacitu) [inline src/resources/views/crud/list.blade.php:41]: Just `print` fixes - looks like we didn't change the classes when moving from Bootstrap 3 to Bootstrap 4.

File: src/resources/views/crud/list.blade.php [hunk 8]
```diff
@@ -74,33 +74,24 @@
...
+                      data-force-export="{{var_export($column['visibleInExport'] ?? true)}}"
                     @endif
                   >
                     {!! $column['label'] !!}
                   </th>
                 @endforeach
 
                 @if ( $crud->buttons()->where('stack', 'line')->count() )
-                  <th data-orderable="false" data-priority="{{ $crud->getActionsColumnPriority() }}" data-visible-in-export="false">{{ trans('backpack::crud.actions') }}</th>
+                  <th data-orderable="false"
```
Comment (tabacitu) [inline src/resources/views/crud/list.blade.php:101]: Just put them on multiple lines.

File: src/resources/views/crud/reorder.blade.php [hunk 9]
```diff
@@ -18,7 +18,7 @@
         <small>{!! $crud->getSubheading() ?? trans('backpack::crud.reorder').' '.$crud->entity_name_plural !!}.</small>
 
         @if ($crud->hasAccess('list'))
-          <small><a href="{{ url($crud->route) }}" class="hidden-print font-sm"><i class="la la-angle-double-left"></i> {{ trans('backpack::crud.back_to_all') }} <span>{{ $crud->entity_name_plural }}</span></a></small>
+          <small><a href="{{ url($crud->route) }}" class="d-print-none font-sm"><i class="la la-angle-double-left"></i> {{ trans('backpack::crud.back_to_all') }} <span>{{ $crud->entity_name_plural }}</span></a></small>
```
Comment (tabacitu) [inline src/resources/views/crud/reorder.blade.php:21]: Just `print` fixes - looks like we didn't change the classes when moving from Bootstrap 3 to Bootstrap 4.
